Implement durable task base class

gherceg opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
https://dimagi.atlassian.net/browse/SAAS-18605

Builds off of https://github.com/dimagi/commcare-hq/pull/37075.

~~This is very much a "first draft" of this feature, and is not yet complete. The main things to consider here are:~~
~- what behavior do we want when updating task records after the task runs successfully? Keep the record around for a few days or just remove it from the table? My understanding is that either way the table churn is equivalent.~
~- Celery task states versus our own task states. I wanted to keep states as close together as possible. This PR follows the same approach that we currently have with custom code that sets the "SENT" state on task results [here](https://github.com/dimagi/commcare-hq/blob/5ac5d947524a87fc191c275d2822b90668fba36d/corehq/celery_monitoring/signals.py#L48-L48).~
~- Currently any failure in apply_async will result in the "PENDING" state. It seems worthwhile to distinguish between OperationalErrors (what is raised when the broker is unavailable) vs other unexpected errors.~

This will become the base class for all tasks, giving us the option to specify `durable=True` on a celery task which will create a TaskRecord object to track the existence of this task prior to it being processed. This enables not losing important tasks that were either never queued in the first place due to network issues, broker issues, etc, or lost by the broker.

Instead of aligning with celery states, there is just a boolean to represent if the task record was sent to the broker or not, and once processed, the record is deleted.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Testing locally and will continue to test on staging.
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Added tests in `corehq/apps/celery/tests/test_durable_task.py`

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
Not particularly applicable for QA since it requires monitoring rows in the db, so I'll test on staging.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Create DurableTask base class

Expects a "durable" attribute to be set on the task object,
and sets a "durable" arg in the task's headers for inspection
after processing the task.

gherceg committed:
Implement post task run updates to task records

millerdev commented on corehq/apps/celery/tests/test_durable_task.py:
```diff
@@ -0,0 +1,161 @@
+        args, kwargs = self.mock_apply_async.call_args
+        with pytest.raises(KeyError):
+            # attempt to access headers -> durable
+            kwargs['headers']['durable']
```
nit
```suggestion
        assert 'durable' not in kwargs['headers']
```

gherceg commented on corehq/apps/celery/durable.py:
```diff
@@ -0,0 +1,58 @@
+    if headers.get('durable', False):
+        record = TaskRecord.objects.get(task_id=task.request.id)
+        record.state = TaskRecordState.from_celery(kwargs.get('state'))
+        record.save()
```
Oh sure. I was specifically thinking of "how many dead tuples would result from this", but was not taking into consideration that we would also be creating a new row. Yeah I think I agree with deleting it. This data model has served its purpose once a task is completed.

AmitPhulera commented on corehq/apps/celery/durable.py:
```diff
@@ -0,0 +1,40 @@
+            TaskRecord.objects.create(
+                task_id=result.task_id if result else None,
+                name=self.name,
+                args=json.dumps(args, cls=CommCareJSONEncoder),
```
I think `CommCareJSONEncoder` would only help with datetime conversions? What would happen if there is a pickled arg that is sent here?

gherceg committed:
Cleaner test assertion

gherceg commented on corehq/apps/celery/tests/test_durable_task.py:
```diff
@@ -0,0 +1,161 @@
+                'properties': {
+                    'field_one': 'value_one',
+                    'field_two': 'value_two',
+                    'created': '2025-10-31T00:00:00.000000Z',
```
Good point. Not handled transparently, or at least not that I'm aware of. I'm a bit stuck as I'm not sure of a way to know whether the datetime string should be converted back to a datetime or not. It is possible that the celery task itself expects the date as a string, so we can't blindly convert all date strings to a datetime object.

gherceg committed:
Enforce json serialization with durable tasks

gherceg committed:
Use kombu's json serializer

This ensures consistency with how task args are serialized, making it
easier to requeue tasks based on the TaskRecord if needed.

gherceg committed:
Improve error message in TaskRecord.error field

gherceg committed:
Support existing TaskRecords for retried tasks

If a task fails during processing and has logic to retry, the TaskRecord
will be preserved and apply_async will be called again. So apply_async
needs to handle this condition to update the existing record rather than
attempt to create a new one, which would conflict on task_id.

millerdev commented on corehq/apps/celery/durable.py:
```diff
@@ -0,0 +1,83 @@
+                args=kombu_json.dumps(args),
+                kwargs=kombu_json.dumps(kwargs),
+                sent=False,
+            )
```
Would this be too performance intensive?
```suggestion
            )
        else:
            assert record.name == self.name, (record.name, self.name)
            assert record.args == kombu_json.dumps(args), (record.args, kombu_json.dumps(args))
            assert record.kwargs == kombu_json.dumps(kwargs), (record.kwargs, kombu_json.dumps(kwargs))
```

gherceg replied:
[f822684](https://github.com/dimagi/commcare-hq/pull/37079/commits/f82268417b2031e5019d12338742cdb9761d658d)

millerdev commented on corehq/apps/celery/durable.py:
```diff
@@ -0,0 +1,83 @@
+
+    task = kwargs.get('task')
+    try:
+        headers = task.request.headers
```
```suggestion
        headers = task.request.headers or {}
```

gherceg replied:
93eed48e12faecaedfb9e22bc3c9f0425c76f82e

millerdev commented on corehq/apps/celery/durable.py:
```diff
@@ -0,0 +1,83 @@
+
+    if headers.get('durable', False):
+        state = kwargs.get('state')
+        if state in celery_states.READY_STATES:
```
Combine if statements with `and`

gherceg replied:
93eed48e12faecaedfb9e22bc3c9f0425c76f82e

millerdev commented on corehq/apps/celery/tests/test_durable_task.py:
```diff
@@ -0,0 +1,175 @@
+        records = TaskRecord.objects.filter(task_id__isnull=True)
+        assert len(records) == 1
+        assert not records[0].sent
+        assert records[0].name == 'corehq.apps.celery.tests.test_durable_task.durable_task'
```
```suggestion
        record, = TaskRecord.objects.filter(task_id__isnull=True)
        assert not record.sent
        assert record.name == 'corehq.apps.celery.tests.test_durable_task.durable_task'
```

gherceg replied:
93eed48e12faecaedfb9e22bc3c9f0425c76f82e

millerdev commented on corehq/apps/celery/tests/test_durable_task.py:
```diff
@@ -0,0 +1,175 @@
+    def test_existing_record_is_updated_for_retries(self):
+        task_id = uuid.uuid4()
+        TaskRecord.objects.create(task_id=task_id, name=durable_task.__name__, args='[]', kwargs='{}', sent=True)
+        durable_task.apply_async(task_id=str(task_id))
```
Maybe add a comment here explaining why this is calling `apply_async` and not `delay`?

gherceg replied:
93eed48e12faecaedfb9e22bc3c9f0425c76f82e

millerdev commented on corehq/apps/celery/tests/test_durable_task.py:
```diff
@@ -0,0 +1,175 @@
+
+@attr.s(auto_attribs=True)
+class MockTask:
+    request: MockRequest
```
Fine print to bottom! haha

millerdev commented on corehq/apps/celery/tests/test_durable_task.py:
```diff
@@ -0,0 +1,175 @@
+            def durable_task_with_pickling(test_id):
+                pass
+
+    def test_existing_record_is_updated_for_retries(self):
```
Move into `TestUpdateTaskRecord`?

gherceg replied:
Added comments to elaborate on this test, but chose not to move it as it functions better with the setup of this test class as opposed to the other.

millerdev reviewed (commented):
Looking great! 🎉

gherceg committed:
Address minor feedback

gherceg committed:
Locally import DurableTask in shared_task.py

gherceg commented on corehq/apps/celery/tests/test_durable_task.py:
```diff
@@ -0,0 +1,175 @@
+        TaskRecord.objects.create(task_id=task_id, name=durable_task.__name__, args='[]', kwargs='{}', sent=True)
+        durable_task.apply_async(task_id=str(task_id))
+        with pytest.raises(TaskRecord.DoesNotExist):
+            TaskRecord.objects.get(task_id=task_id)
```
Correct. I forgot to add a comment in 93eed48e12faecaedfb9e22bc3c9f0425c76f82e so I'll add that.

gherceg replied:
[52b57e5](https://github.com/dimagi/commcare-hq/pull/37079/commits/52b57e5302eef9b7fec8a55fa13cd3d6ee6535e4)

gherceg commented on corehq/apps/celery/durable.py:
```diff
@@ -0,0 +1,83 @@
+                )
+
+    def apply_async(self, args=None, kwargs=None, **opts):
+        from corehq.apps.celery.models import TaskRecord
```
7011f5f1178d5c4cf43766b7240ee257b1efebdf

gherceg committed:
Serialize args and kwargs for existing records

New args and kwargs can be specified when retrying a celery task, so
just update the record to be safe.

gherceg committed:
Add comments for clarity

gherceg committed:
Ensure durable=False behaves as expected

millerdev commented on corehq/apps/celery/shared_task.py:
Seems like this is no longer true.

gherceg committed:
Update docs for the HQ task decorator

gherceg committed:
Handle race where TaskRecord is created by worker

Since the original apply_async call does not create a TaskRecord until
after calling super().apply_async, it is possible for a worker to pick
up the task and create a TaskRecord when retrying the task. To mitigate
this, we use update_or_create instead, making no change if the
subsequent call to apply_async won the race condition since that
information is more up to date anyway.

gherceg committed:
Reset sent and error attributes on retries

Added tests to prove that sent and error can be updated on subsequent
attempts/retries. No code change was necessary for 'sent' since it was
included in defaults, but we needed to include 'error' in defaults as
well.

AmitPhulera reviewed (dismissed):
Looking great 🥇

millerdev commented on corehq/apps/celery/durable.py:
```diff
@@ -0,0 +1,72 @@
+            update_defaults = defaults if opts.get('task_id', None) else {}
+            TaskRecord.objects.update_or_create(
+                task_id=task_id, create_defaults=defaults, defaults=update_defaults
+            )
```
The Django docs are a bit unclear, but maybe `unique_fields=['task_id'],` should be changed to `unique_fields=['task_id'] if is_retry else None,`?

gherceg replied:
Yeah just realized a bug in the existing code is that when `task_id` is None (so failed to send to the broker), a TaskRecord object is not created because the current logic only creates a new TaskRecord if the task_id is unique. It is tempting to just check when task_id is None and go with a standard create in that case.

The downsize of `bulk_create` is that it doesn't trigger auto_now and auto_now_add since the `save` method isn't called, so that would need to be manually handled here.

gherceg replied:
Ok based on how this behaves in tests I think I was wrong about auto_now and auto_now_add not behaving as expected with `bulk_create`, so nevermind!

gherceg committed:
Use after_return handler instead of task_postrun

This is a handler that gets called directly on the Task class rather
than a signal that we need to connect to, and can take advantage of the
self.durable attribute directly available on the class rather than
relying on headers to pass around the durable flag.

I observed issues with the durable flag not being available in the
headers dict in task_postrun after the task was successful on staging,
which meant the TaskRecord was not being cleaned up.

gherceg committed:
Use bulk_create to get support for upserts

bulk_create is the only way to ensure postgres performs an upsert,
combining an insert and update statement into one operation.

gherceg committed:
Update docstrings

gherceg commented on corehq/apps/celery/shared_task.py:
```diff
@@ -6,23 +6,29 @@
+
+    - enforces the default task serializer as JSON, which is needed until
+    https://github.com/celery/celery/issues/6759 is resolved, at which point CELERY_TASK_SERIALIZER
+    can be set back to json.
```
[0873882](https://github.com/dimagi/commcare-hq/pull/37079/commits/08738825567ab11cc88a2386831ea96f286f3a10)

millerdev commented on corehq/apps/celery/shared_task.py:
```diff
@@ -6,23 +6,30 @@
+    record of incomplete tasks
+    - enforces the default task serializer as JSON, which is needed until
+    https://github.com/celery/celery/issues/6759 is resolved, at which point
+    CELERY_TASK_SERIALIZER can be set back to json.
```
nit
```suggestion
    - Adds "durable" task functionality that optionally creates a permanent
      record of incomplete tasks
    - enforces the default task serializer as JSON, which is needed until
      https://github.com/celery/celery/issues/6759 is resolved, at which point
      CELERY_TASK_SERIALIZER can be set back to json.
```

gherceg commented on corehq/apps/celery/durable.py:
```diff
@@ -53,19 +51,20 @@ def apply_async(self, args=None, kwargs=None, **opts):
                 task_id=task_id, create_defaults=defaults, defaults=update_defaults
             )
 
+    def after_return(self, state, retval, task_id, *args):
```
> Is this called for ignore_result=True tasks? Is that even compatible with durable tasks?

When ignore_result is True, _the task state is not stored_, but this isn't used by durable tasks. The only thing durable task relies on is the task id being returned after calling `apply_async` which it is, and the `after_return` handler being called once the task is processed by the worker, which it is.

> What calls this? Is it the celery worker? I can't really imagine any other place that could call it reliably, but maybe I'm missing something.

Yeah the worker right after processing the task.  We can define a number of [handlers](https://docs.celeryq.dev/en/stable/userguide/tasks.html#handlers) on the task object, and when the worker runs the task, it will check [if this handler exists](https://github.com/celery/celery/blob/afb67303ae95c9af499412940046a6c0a3b46a51/celery/app/trace.py#L383-L384), and [if so call it](https://github.com/celery/celery/blob/afb67303ae95c9af499412940046a6c0a3b46a51/celery/app/trace.py#L575-L579) when the task completes.

gherceg committed:
Merge branch 'gh/task-record' into gh/durable-task

gherceg committed:
Merge branch 'gh/task-record' into gh/durable-task

gherceg committed:
Create TaskRecord prior to calling apply_async

This avoids the race condition of a celery worker completing and
attempting to delete a TaskRecord before the producer has created it.

Additionally, the sent field is no longer needed since it is redundant
of the error column. And all durable tasks will be responsible for
creating their own task_id. The TaskRecord table has a default of
uuid.uuid4 for the task_id, which is the primary key as well.

gherceg committed:
Explain how durable is an attr on self

We could either add a default value of False for durable in our task
decorator code, or bury that logic in the DurableTask object itself,
which is what I chose to do here.

https://github.com/celery/celery/blob/63c1910221d9fb8c1931ff62c80c4c37232b2d58/celery/app/base.py#L586-L595

gherceg committed:
Fix docstring indentation

gherceg committed:
Add detail to when attributes are added

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

gherceg committed:
Merge branch 'master' into gh/durable-task

gherceg committed:
Run make serializer-pickle-files.lock

Allows usage of serializer='pickle' in a test that verifies pickling is
not compatible with durable tasks
